sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedPartitions.scala

```diff
@@ -0,0 +1,281 @@
+    mapOutputTracker.shuffleStatuses.get(shuffleId).
+      get.mapStatuses.map{_.getSizeForBlock(partitionId)}
+  }
+
+  /*
+  * Split the mappers based on the map size of specific skewed reduce partitionId.
+  */
+  def splitMappersBasedDataSize(mapPartitionSize: Array[Long], numMappers: Int): Array[Int] = {
+    val advisoryTargetPostShuffleInputSize = conf.targetPostShuffleInputSize
+    val partitionStartIndices = ArrayBuffer[Int]()
+    var i = 0
+    var postMapPartitionSize: Long = mapPartitionSize(i)
+    partitionStartIndices += i
+    while (i < numMappers && i + 1 < numMappers) {
+      val nextIndex = if (i + 1 < numMappers) {
+        i + 1
+      } else numMappers -1
+
+      if (postMapPartitionSize + mapPartitionSize(nextIndex) > advisoryTargetPostShuffleInputSize) {
+        postMapPartitionSize = mapPartitionSize(nextIndex)
+        partitionStartIndices += nextIndex
+      } else {
+        postMapPartitionSize += mapPartitionSize(nextIndex)
+      }
+      i += 1
+    }
+    partitionStartIndices.toArray
+  }
+
+  /**
+   * We split the partition into several splits. Each split reads the data from several map outputs
+   * ranging from startMapId to endMapId(exclusive). This method calculates the split number and
+   * the startMapId for all splits.
+   */
+  private def estimateMapIdStartIndices(
+    stage: QueryStageExec,
+    partitionId: Int,
+    medianSize: Long): Array[Int] = {
+    val dependency = getShuffleStage(stage).plan.shuffleDependency
+    val shuffleId = dependency.shuffleHandle.shuffleId
+    val mapSize = getMapSizeForSpecificPartition(partitionId, shuffleId)
+    val numMappers = dependency.rdd.partitions.length
+    splitMappersBasedDataSize(mapSize, numMappers)
+  }
+
+  private def getShuffleStage(queryStage: QueryStageExec): ShuffleQueryStageExec = {
+    queryStage match {
+      case stage: ShuffleQueryStageExec => stage
+      case ReusedQueryStageExec(_, stage: ShuffleQueryStageExec, _) => stage
+    }
+  }
+
+  private def getStatistics(queryStage: QueryStageExec): MapOutputStatistics = {
+    val shuffleStage = queryStage match {
+      case stage: ShuffleQueryStageExec => stage
+      case ReusedQueryStageExec(_, stage: ShuffleQueryStageExec, _) => stage
+    }
+    val metrics = shuffleStage.mapOutputStatisticsFuture
+    assert(metrics.isCompleted, "ShuffleQueryStageExec should already be ready")
+    ThreadUtils.awaitResult(metrics, Duration.Zero)
+  }
+
+  /**
+   * Base optimization support check: the join type is supported and plan statistics is available.
+   * Note that for some join types(like left outer), whether a certain partition can be optimized
+   * also depends on the filed isSkewAndSupportsSplit.
+   */
+  private def supportOptimization(
+    joinType: JoinType,
+    leftStage: QueryStageExec,
+    rightStage: QueryStageExec): Boolean = {
+    val joinTypeSupported = supportedJoinTypes.contains(joinType)
```

Comment:
do we need to check join type here? We need to check it again later when we know which side is skewed.

Reply:
If we check the join type here, we may can directly skip the following optimization when encounter "Full Outer" type ?